File: src/application/services/webhook_service.py
from __future__ import annotations

import logging
import secrets

from src.domain.model.tenant.webhook import Webhook
from src.domain.ports.repositories.webhook_repository import WebhookRepository

logger = logging.getLogger(__name__)


class WebhookService:
    def __init__(self, webhook_repo: WebhookRepository) -> None:
        self._webhook_repo = webhook_repo

    async def create_webhook(
        self,
        tenant_id: str,
        name: str,
        url: str,
        events: list[str],
        is_active: bool = True,
    ) -> Webhook:
        secret = f"whsec_{secrets.token_hex(32)}"

        webhook = Webhook(
            tenant_id=tenant_id,
            name=name,
            url=url,
            secret=secret,
            events=events,
            is_active=is_active,
        )
        await self._webhook_repo.save(webhook)
        logger.info(f"Created webhook {webhook.id} for tenant {tenant_id}")
        return webhook

    async def get_webhook(self, webhook_id: str) -> Webhook | None:
        return await self._webhook_repo.get_by_id(webhook_id)

    async def list_webhooks(self, tenant_id: str) -> list[Webhook]:
        return await self._webhook_repo.list_by_tenant(tenant_id)

    async def update_webhook(
        self,
        webhook_id: str,
        name: str,
        url: str,
        events: list[str],
        is_active: bool,
    ) -> Webhook:
        webhook = await self._webhook_repo.get_by_id(webhook_id)
        if not webhook:
            raise ValueError(f"Webhook {webhook_id} not found")

        webhook.update(
            name=name,
            url=url,
            secret=webhook.secret,
            events=events,
            is_active=is_active,
        )
        await self._webhook_repo.save(webhook)
        logger.info(f"Updated webhook {webhook.id}")
        return webhook

    async def delete_webhook(self, webhook_id: str) -> bool:
        deleted = await self._webhook_repo.delete_by_id(webhook_id)
        if deleted:
            logger.info(f"Deleted webhook {webhook_id}")
        return deleted

File: web/src/services/webhookService.ts
import { httpClient } from './client/httpClient';

export interface Webhook {
  id: string;
  tenant_id: string;
  name: string;
  url: string;
  secret: string | null;
  events: string[];
  is_active: boolean;
  created_at: string | null;
  updated_at: string | null;
}

export interface WebhookCreateRequest {
  name: string;
  url: string;
  events: string[];
  is_active?: boolean;
}

export interface WebhookUpdateRequest {
  name: string;
  url: string;
  events: string[];
  is_active: boolean;
}

export const webhookService = {
  async listWebhooks(tenantId: string): Promise<Webhook[]> {
    return httpClient.get<Webhook[]>(`/tenant-webhooks/${tenantId}`);
  },

  async createWebhook(tenantId: string, data: WebhookCreateRequest): Promise<Webhook> {
    return httpClient.post<Webhook>(`/tenant-webhooks/${tenantId}`, data);
  },

  async updateWebhook(webhookId: string, data: WebhookUpdateRequest): Promise<Webhook> {
    return httpClient.put<Webhook>(`/tenant-webhooks/${webhookId}`, data);
  },

  async deleteWebhook(webhookId: string): Promise<void> {
    await httpClient.delete(`/tenant-webhooks/${webhookId}`);
  },
};
